### PR TITLE
lmnn cookbook

### PR DESCRIPTION
**Pull request by Saurabh7**

### CODE REVIEW

**Comment by Saurabh7** [comment]
Will add a reference here.

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
+=================================
+Largest Margin Nearest Neighbours
+=================================
+
+Largest Margin Nearest Neighbours is a metric learning algorithm.  It learns a metric that can be used with the K-Nearest neigbour algorithm.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: link to knn cookbook maybe?

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+The Mahalanobis distance metric which is obtained is computed for  samples :math:`x_i` and :math:`x_j` as:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: Can you say that this distance is one example that can be used?

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+.. math::
+
+	( (\vec{x_i} - \vec{x_j})^T \bf{M} (\vec{x_i} - \vec{x_j}) )
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: No need to provide the math, just sgclass

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+We create an instance of :sgclass:`CLMNN` with parameter k (nearest neighbours).
+
+.. sgexample:: lmnn.sg:create_instance
+
+Next we train the LMNN algorithm and get the learned metric.
+
+.. sgexample:: lmnn.sg:train_metric
+
+Then we run the train KNN algorithm using the learned metric and apply it to test data, which here gives :sgclass:`CMulticlassLabels`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: sgclass knn

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 5]
```diff
@@ -0,0 +1,47 @@
+=================================
+Largest Margin Nearest Neighbours
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: please dont abbreviate filenames. Call it "largest_margin_nearest_neighbours.rst"

**Comment by karlnapf** [comment]
integration test data

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+-------
+Example
+-------
+
+Imagine we have files with training and test data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CMulticlassLabels` as
+
+.. sgexample:: lmnn.sg:create_features
+
+We create an instance of :sgclass:`CLMNN` with parameter k (nearest neighbours).
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: dont mention parameter names in the text (they might change),
just say that you provide the number of neighbours as parameter

File: doc/cookbook/source/examples/classifier/lmnn.rst [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+.. sgexample:: lmnn.sg:evaluate_accuracy
+
+
+----------
+References
+----------
+:wiki:`Large_margin_nearest_neighbor`
+
+.. bibliography:: ../../references.bib
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: can you add a bibtex

File: examples/meta/src/classifier/lmnn.sg [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+#![train_metric]
+
+#![train_and_apply]
+KNN knn(k, lmnn_distance,labels_train)
+knn.train()
+MulticlassLabels labels_predict = knn.apply_multiclass(features_test)	
+#![train_and_apply]
+
+#![evaluate_accuracy]	
+MulticlassAccuracy eval()
```
Comment (karlnapf) [inline examples/meta/src/classifier/lmnn.sg]: please dont call this "eval"

File: doc/cookbook/source/examples/classifier/lmnn.rst (hunk 2 above)
Comment (Saurabh7) [inline doc/cookbook/source/examples/classifier/lmnn.rst]: I think LMNN always returns `CustomMahalanobisDistance`. http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CLMNN.html#a9f96703addfc52317b1f2bb97e20a7ee

I have mentioned it as sgclass now.

**Comment by Saurabh7** [comment]
updated. added data here https://github.com/shogun-toolbox/shogun-data/pull/114

**Comment by karlnapf** [comment]
ready, pls squash and add the merged data

File: doc/cookbook/source/examples/multiclass_classifier/large_margin_nearest_neighbours.rst [hunk 9]
```diff
@@ -0,0 +1,43 @@
+=================================
```
Comment (karlnapf) [inline doc/cookbook/source/examples/multiclass_classifier/large_margin_nearest_neighbours.rst]: can you align this?

File: doc/cookbook/source/references.bib [hunk 10]
```diff
@@ -109,3 +109,12 @@ @article{Breiman2001
 volume={45},
 pages={5--32}
 }
+@article{weinberger2009distance,
+  title={Distance metric learning for large margin nearest neighbor classification},
+  author={Weinberger, Kilian Q and Saul, Lawrence K},
```
Comment (karlnapf) [inline doc/cookbook/source/references.bib]: pls abbreviate first names

**Comment by vigsterkr** [comment]
ping ping ping let's finish and merge!

**Comment by Saurabh7** [comment]
Updated, need to check travis.

**Pull request merged**